dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+  should for the others
+* Same issue with `FetchHistory*` as `ListWorkflows` on client - no `IAsyncEnumerator` on really old .NET versions, but
+  we're ok with that
+
+## Worker
+
+Usage:
+
+```csharp
+using Temporal.Client;
+using Temporal.Worker;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Build worker options
+var options = new TemporalWorkerOptions("task queue");
+options.AddActivities(new GreetingActivities());
+
+// Run until Ctrl+C
+var worker = new TemporalWorker(client, options);
+var cts = new CancellationTokenSource();
+Console.CancelKeyPress += (sender, eventArgs) => {
+    eventArgs.Cancel = true;
+    cts.Cancel();
+};
+await worker.ExecuteAsync(cts.Token);
+```
+
+Looks like:
+
+```csharp
+namespace Temporal.Worker;
+
+using Temporal.Activity;
+
+public class TemporalWorkerOptions
+{
+    public TemporalWorkerOptions() { /*...*/ }
+    public TemporalWorkerOptions(string taskQueue) { /*...*/ }
+
+    public string? TaskQueue { get; set; }
+
+    public List<ActivityDefinition> ActivityDefinitions { get; set; } = new();
+
+    public void AddActivities(params object[] activities) { /*...*/ }
+
+    public void AddActivityDefinition(ActivityDefinition definition) { /*...*/ }
+
+    // Many more options not listed
+}
+
+public class TemporalWorker
+{
+    public TemporalWorker(IWorkerClient client, TemporalWorkerOptions options) { /*...*/ }
+
+    public Task ExecuteAsync(CancellationToken stoppingToken) { /*...*/ }
+}
+```
+
+Notes:
+
+* Like other options, technically all worker options are optional at compile time but we fail at runtime if task queue
+  is not set
+* Activity definitions are the underlying type of the activity collection (they have just a name and method to call) but
+  most users will call `AddActivities`
+* `AddActivities` intentionally take types or instances
+  * For types, all `[Activity`s must be on static methods, no instance methods allowed
+    * It was decided to do `AddActivities(typeof(MyStaticClass))` for all static method things than to have
+    `AddActivities<MyStaticClass>()` overload. All-static activities may not be that common and adding overloads is
+    unnecessary.
+  * For instances, `[Activity`s can be on static or instance methods
+  * Inheritance rules apply from the earlier activity definition section
+  * Users can use the lower-level `AddActivityDefinition` but we left this name unwieldy on purpose, we prefer
+    `AddActivities`
+* Since workflows are not currently implemented, at least one activity is required
+* Activities can be sync or async
+  * For future `Workflow.ExecuteActivity` uses, we have confirmed .NET can disambiguate in an overload between `Task`
+    return type and generic return type with no constraints
+  * Async activities are executed using the normal scheduler
+    * No need yet for a custom scheduler
+  * Sync activities are executed with `ThreadPool`
```

Comment:
Is this a fixed size thread pool? How does this work with the worker limitations of `maxConcurrent(Local)Activities`?
Can user get themselves into a situation where the worker polls on an activity and there's no free slot on the pool to execute it?

Reply:
> Is this a fixed size thread pool?

This is a global .NET thing: https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadpool. There is not even a way to make a custom thread pool in .NET. It's kinda weird but very common for .NET developers to use this.

> How does this work with the worker limitations of maxConcurrent(Local)Activities?
Can user get themselves into a situation where the worker polls on an activity and there's no free slot on the pool to execute it?

It'll just pool up work until a thread is open for use. So they could hit heartbeat timeouts and, if I don't have a thread set aside for workflows (I probably won't) task timeouts there. I can't really think of how to improve this. They should set `SetMinThreads` globally probably, but if everyone properly uses async, we shouldn't make them (those are non-blocking but still use the same thread pool).

You might think you could warn a user with ".NET thread pool has fewer threads than concurrent activity limit", but again with async that's unfair. And people may be totally ok with having a received activity wait on an available thread (workflows less so, will discuss when we get there). And .NET'ters are used to this limitation.

I am open to alternatives, but short of building custom thread pools the user can pass around, there isn't much we can do besides docs that I can think of.